x/stablestake/keeper/debt.go

```diff
@@ -33,6 +34,99 @@ func (k Keeper) UpdateInterestStackedByAddress(ctx sdk.Context, addr sdk.AccAddr
 	return debt
 }
 
+func (k Keeper) SetInterest(ctx sdk.Context, block uint64, interest types.InterestBlock) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.InterestPrefixKey)
+	if store.Has(sdk.Uint64ToBigEndian(block - 1)) {
+		lastBlock := types.InterestBlock{}
+		bz := store.Get(sdk.Uint64ToBigEndian(block - 1))
+		k.cdc.MustUnmarshal(bz, &lastBlock)
+		interest.InterestRate = interest.InterestRate.Add(lastBlock.InterestRate)
+
+		bz = k.cdc.MustMarshal(&interest)
+		store.Set(sdk.Uint64ToBigEndian(block), bz)
+	} else {
+		bz := k.cdc.MustMarshal(&interest)
+		store.Set(sdk.Uint64ToBigEndian(block), bz)
+	}
+}
+
+func (k Keeper) GetAllInterest(ctx sdk.Context) []types.InterestBlock {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.InterestPrefixKey)
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+	defer iterator.Close()
+
+	interests := []types.InterestBlock{}
+	for ; iterator.Valid(); iterator.Next() {
+		interest := types.InterestBlock{}
+		k.cdc.MustUnmarshal(iterator.Value(), &interest)
+
+		interests = append(interests, interest)
+	}
+	return interests
+}
+
+func (k Keeper) GetInterest(ctx sdk.Context, startBlock uint64, startTime uint64, borrowed sdk.Dec) sdk.Int {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.InterestPrefixKey)
+	currentBlockKey := sdk.Uint64ToBigEndian(uint64(ctx.BlockHeight()))
+	startBlockKey := sdk.Uint64ToBigEndian(startBlock)
+
+	// note: exclude start block
+	if store.Has(startBlockKey) && store.Has(currentBlockKey) && startBlock != uint64(ctx.BlockHeight()) {
+		bz := store.Get(startBlockKey)
+		startInterestBlock := types.InterestBlock{}
+		k.cdc.MustUnmarshal(bz, &startInterestBlock)
+
+		bz = store.Get(currentBlockKey)
+		endInterestBlock := types.InterestBlock{}
+		k.cdc.MustUnmarshal(bz, &endInterestBlock)
+
+		totalInterest := endInterestBlock.InterestRate.Sub(startInterestBlock.InterestRate)
+		numberOfBlocks := ctx.BlockHeight() - int64(startBlock)
+
+		newInterest := borrowed.
+			Mul(totalInterest).
+			Mul(sdk.NewDec(ctx.BlockTime().Unix() - int64(startTime))).
+			Quo(sdk.NewDec(numberOfBlocks)).
+			Quo(sdk.NewDec(86400 * 365)).
+			RoundInt()
+		return newInterest
+	}
+
+	if !store.Has(startBlockKey) && store.Has(currentBlockKey) {
+		iterator := sdk.KVStorePrefixIterator(store, nil)
+		defer iterator.Close()
+
+		firstStoredBlock := uint64(0)
+		if iterator.Valid() {
+			interestBlock := types.InterestBlock{}
+			firstStoredBlock = sdk.BigEndianToUint64(iterator.Key())
+			k.cdc.MustUnmarshal(iterator.Value(), &interestBlock)
+		}
+		if firstStoredBlock > startBlock {
+			bz := store.Get(currentBlockKey)
+			endInterestBlock := types.InterestBlock{}
+			k.cdc.MustUnmarshal(bz, &endInterestBlock)
+
+			totalInterest := endInterestBlock.InterestRate
+			numberOfBlocks := ctx.BlockHeight() - int64(startBlock) + 1
+
+			newInterest := borrowed.Mul(totalInterest).
+				Mul(sdk.NewDec(ctx.BlockTime().Unix() - int64(startTime))).
```

Comment:
same as above

Reply:
no, we can't do that because we don't have startBlock in this case

Reply:
We can then fetch it outside the if blocks, and then use it the both